First draft for Lyrion Music Server (LMS, Logitech Music Server) app

hartzell opened the pull request:
First draft for an app for the Lyrion Music Server (used to be Logitech Music Server).

Bits stolen from flaresolverr (extant app) and Parsonlabs Music (MR), chosen at random.

 With the values in the `basic-values.yaml` test inputs file, this seems to work:

     .github/scripts/ci.py --app lms --train community --test-file basic-values.yaml

If I hack those answers up a bit and mount a real tree full of music into the volume then I can even index music.

I've marked this ready for review, but ...  I need to get some feedback from the ixSystems folks.

TODO:
- screenshots

FEEDBACK requests:
- Other apps use perm containers to do things.  Any docs about what/how/why/when?

stavros-k commented:
Hello, I'll take a look shortly.

Regarding the permissions, this is only needed when the container runs rootless.
In this case the container runs as root, and the sub processes of it run rootless. Still the pid 1 runs as root.
Eg if you see the upstream dockerfile https://github.com/LMS-Community/slimserver-platforms/blob/9c80a1dc89d8ec42dad5a7b28321c6db172d911b/Docker/start-container.sh#L5-L10

at startup it will create the user/group based on the PUID/PGID and then fix the permissions itself.

stavros-k commented on ix-dev/community/lms/app.yaml:
```diff
@@ -0,0 +1,32 @@
+lib_version: 2.1.7
+lib_version_hash: c4c34d5b8aa0e8c3016b7cb3ee8a78ef09d4ffe80f34eb35107810656a43a615
+maintainers:
+- email: hartzell@alerce.com
+  name: George Hartzell
+  url: https://github.com/hartzell
```
Please use the ix systems as maintainer here.

stavros-k commented on ix-dev/community/lms/app.yaml:
```diff
@@ -0,0 +1,32 @@
+- email: hartzell@alerce.com
+  name: George Hartzell
+  url: https://github.com/hartzell
+name: lms
```
Lets rename it to `lyrion-music-server` (also remember to rename the directory)

stavros-k commented on ix-dev/community/lms/app.yaml:
```diff
@@ -0,0 +1,32 @@
+  url: https://github.com/hartzell
+name: lms
+run_as_context:
+- description: LMS runs as any non-root user
+  gid: 568
+  group_name: lms
+  uid: 568
+  user_name: lms
```
This should say root, as the main pid is running as root

stavros-k commented on ix-dev/community/lms/ix_values.yaml:
```diff
@@ -0,0 +1,10 @@
+
+consts:
+  lms_container_name: lms
+  run_as_user: 568
+  run_as_group: 568
```
No need to add those as constants, they are configurable

stavros-k commented on ix-dev/community/lms/questions.yaml:
```diff
@@ -0,0 +1,488 @@
+            required: true
+            $ref:
+              - definitions/port
+
```
we need a run as section here

stavros-k commented on ix-dev/community/lms/templates/docker-compose.yaml:
```diff
@@ -0,0 +1,26 @@
+  {% do c1.add_storage(store.mount_path, store) %}
+{% endfor %}
+
+{% do c1.environment.add_env("PUID", values.consts.run_as_user) %}
+{% do c1.environment.add_env("PGID", values.consts.run_as_group) %}
```
once you add the run_as section in both the test file and the questions, you wont need those. it will pick it up automatically

stavros-k replied:
also add the line to add additional env variables.

stavros-k commented on ix-dev/community/lms/templates/docker-compose.yaml:
```diff
@@ -0,0 +1,26 @@
+{% do c1.environment.add_env("PUID", values.consts.run_as_user) %}
+{% do c1.environment.add_env("PGID", values.consts.run_as_group) %}
+
+{% do c1.healthcheck.set_test("curl", {"port": values.network.web_port, "path": "/"}) %}
```
```suggestion
{% do c1.healthcheck.set_test("curl", {"port": values.network.web_port}) %}
```
`/` is the default

stavros-k commented on ix-dev/community/lms/templates/test_values/basic-values.yaml:
```diff
@@ -0,0 +1,35 @@
+    cpus: 2.0
+    memory: 4096
+
+music:
```
```suggestion
lms:
```

hartzell commented:
Thank you!

stavros-k commented on ix-dev/community/lms/templates/docker-compose.yaml:
```diff
@@ -0,0 +1,26 @@
+{% endfor %}
+
+{% do c1.environment.add_env("PUID", values.consts.run_as_user) %}
+{% do c1.environment.add_env("PGID", values.consts.run_as_group) %}
```
```suggestion
{% do c1.environment.add_env("PGID", values.consts.run_as_group) %}
{% do c1.environment.add_env("HTTP_PORT", values.network.web_port.port_number) %}
```

stavros-k replied:
<img width="1058" alt="image" src="https://github.com/user-attachments/assets/aeb19b8f-3193-4d06-8881-097f791bb940" />

The cli port is gonna be a problem at some point tho.

hartzell replied:
Added your suggestion.  Scratching head a bit about the CLI port "problem", will revisit this comment after working through the other PORT bits below.

stavros-k commented on ix-dev/community/lms/questions.yaml:
```diff
@@ -0,0 +1,488 @@
+    schema:
+      type: dict
+      attrs:
+        - variable: web_port
+          label: WebUI port
+          description: The port the web UI (default 9000).
+          schema:
+            type: int
+            default: 9000
+            required: true
+            $ref:
+              - definitions/port
```
```suggestion
        - variable: web_port
          label: WebUI Port
          schema:
            type: dict
            attrs:
              - variable: bind_mode
                label: Port Bind Mode
                description: |
                  The port bind mode.</br>
                  - Publish: The port will be published on the host for external access.</br>
                  - Expose: The port will be exposed for inter-container communication.</br>
                  - None: The port will not be exposed or published.</br>
                  Note: If the Dockerfile defines an EXPOSE directive,
                  the port will still be exposed for inter-container communication regardless of this setting.
                schema:
                  type: string
                  default: "published"
                  enum:
                    - value: "published"
                      description: Publish port on the host for external access
                    - value: "exposed"
                      description: Expose port for inter-container communication
                    - value: ""
                      description: None
              - variable: port_number
                label: Port Number
                schema:
                  type: int
                  show_if: [["bind_mode", "!=", ""]]
                  default: 31101
                  required: true
                  $ref:
                    - definitions/port
```

This is the new format for port

hartzell replied:
Done, but used 9000 as default, not the 31101 in your change.  Am I mis-understanding?

stavros-k replied:
I prefer to have unique ports across apps, to avoid conflicts.

hartzell replied:
If I understand correctly, something else already has 9000 so you're using something else here, but nothing else has 9090 so you used that for the cli port.
If that's correct, I'll change 9000 -> 31101.

Is there a place to warn people that, unless they take action, LMS will end up running on a non-traditional port?

stavros-k commented on ix-dev/community/lms/questions.yaml:
```diff
@@ -0,0 +1,488 @@
+            required: true
+            $ref:
+              - definitions/port
+        - variable: cli_port
+          label: CLI port
+          description: The port for the CLI (default 9090).
+          schema:
+            type: int
+            default: 9090
+            required: true
+            $ref:
+              - definitions/port
```
```suggestion
        - variable: cli_port
          label: CLI port
          schema:
            type: dict
            attrs:
              - variable: bind_mode
                label: Port Bind Mode
                description: |
                  The port bind mode.</br>
                  - Publish: The port will be published on the host for external access.</br>
                  - Expose: The port will be exposed for inter-container communication.</br>
                  - None: The port will not be exposed or published.</br>
                  Note: If the Dockerfile defines an EXPOSE directive,
                  the port will still be exposed for inter-container communication regardless of this setting.
                schema:
                  type: string
                  default: "published"
                  enum:
                    - value: "published"
                      description: Publish port on the host for external access
                    - value: "exposed"
                      description: Expose port for inter-container communication
                    - value: ""
                      description: None
              - variable: port_number
                label: Port Number
                schema:
                  type: int
                  show_if: [["bind_mode", "==", "published"]]
                  default: 9090
                  required: true
                  $ref:
                    - definitions/port
```

stavros-k commented on ix-dev/community/lms/templates/docker-compose.yaml:
```diff
@@ -0,0 +1,26 @@
+
+{% do c1.add_caps(["CHOWN", "SETGID", "SETUID"]) %}
+
+{% do c1.ports.add_port(9000, values.network.web_port) %}
+{% do c1.ports.add_port(9090, values.network.cli_port) %}
+{% do c1.ports.add_port(3483, 3483) %}
+{% do c1.ports.add_port(3483, 3483, {"protocol": "udp"}) %}
```
```suggestion
{% do c1.add_port(values.network.web_port) %}
{# TODO: We might have to expose the container port as well on ui #}
{% do c1.add_port(values.network.cli_port, {"container_port": 9090}) %}
{# FIXME: those should be exposed on UI too, and use the c1.add_port() #}
{% do c1.ports.add_port(3483, 3483) %}
{% do c1.ports.add_port(3483, 3483, {"protocol": "udp"}) %}
```
Left couple todo's. I'll check them later

hartzell replied:
Done, left the TODO/FIXME's in place

stavros-k commented on ix-dev/community/lms/app.yaml:
```diff
@@ -0,0 +1,32 @@
+app_version: 9.0.0
+capabilities: []
+categories:
```
currently ui supports only 1 category

hartzell replied:
done, went with "media"

stavros-k commented on ix-dev/community/lms/app.yaml:
```diff
@@ -0,0 +1,32 @@
+app_version: 9.0.0
+capabilities: []
```
Missing the capabilities here (chown, etc)

hartzell replied:
done

stavros-k reviewed (changes requested):
Thanks for the contribution! Left few comments.!
Overall looks fine, just small nits here and there.

Once those are fixed, I'll take a closer look as well!
Thansk!

hartzell commented:
@stavros-k -- I think I've addressed your feedback (assuming I did the right thing with the web port bit I left unresolved above).  I can run the server using the generated docker-compose.yaml and it seems to work (haven't fed it any real streams or music yet though).

hartzell commented:
For screenshots, how about

- https://lyrion.org/assets/screenshot.png

stavros-k commented on ix-dev/community/lyrion-music-server/app.yaml:
```diff
@@ -0,0 +1,37 @@
+  url: https://www.truenas.com/
+name: lyrion-music-server
+run_as_context:
+- description: Main pid runs as root
```
```suggestion
- description: Lyrion Music Server runs as root user.
```

stavros-k commented on ix-dev/community/lyrion-music-server/ix_values.yaml:
```diff
@@ -0,0 +1,9 @@
+
+consts:
+  lms_container_name: lms
+
```
```suggestion
```

hartzell replied:
done

stavros-k commented on ix-dev/community/lyrion-music-server/templates/test_values/basic-values.yaml:
```diff
@@ -0,0 +1,44 @@
+  group: 568
+
+ix_volumes:
+  config: /mnt/storage/lms/config
+  music: /mnt/storage/lms/music
+  playlist: /mnt/storage/lms/playlist
```
```suggestion
  config: /opt/tests/mnt/storage/lms/config
  music: /opt/tests/mnt/storage/lms/music
  playlist: /opt/tests/mnt/storage/lms/playlist
```

stavros-k replied:
Mostly to keep it consistent with the rest.
I dont want someone to accidentaly run the tests and files in their /mnt/storage are getting written/overwritten

hartzell replied:
Is there a standard way to get `/opt/tests` set up on a TrueNAS Scale system?  `/opt` is on the boot-pool, which is read-only.  I've seen references to remounting it r/w, but also warnings that that's not supported (might lead Support to ignoring questions).

stavros-k replied:
Those paths are picked for local development. ie your laptop/pc
Specifically the /opt/tests are picked because macOS whitelists those by default
(https://github.com/truenas/apps/blob/master/CONTRIBUTIONS.md#appyaml)

hartzell replied:
Yep, I saw that.  But, I'm not doing this work on my laptop, I'm doing it on my TrueNAS scale system.  Those paths are particularly problematic on a Scale box.

I'd happily just carve off a data set from my storage pool, but I don't see any supported way to mount it on /opt/tests.  Trying to to go wandering off into the forest if I can help it.

Any Scale suggestions?

stavros-k replied:
Developing apps on scale is not really easy. Well at least I don't find it easy.
Changing the paths during dev is fine. But I think most people will use their computers to do dev.
Hence insisting on keeping the paths under /opt/tests/

stavros-k commented on ix-dev/community/lyrion-music-server/questions.yaml:
```diff
@@ -0,0 +1,560 @@
+      required: true
+      $ref:
+        - definitions/timezone
+
```
missing the `variable: lms` section with the addtitional envs section

hartzell replied:
done.

stavros-k commented on ix-dev/community/lyrion-music-server/questions.yaml:
```diff
@@ -0,0 +1,560 @@
+groups:
+  - name: LMS Configuration
+    description: Configure Lyrion Music Server
+  - name: Network Configuration
```
Missing the user group here

hartzell replied:
done

stavros-k reviewed (commented):
Looks mostly good. Will check the port thingies in a bit and update the url for the images

stavros-k commented:
> For screenshots, how about
> 
>     * https://lyrion.org/assets/screenshot.png
> 
> 
> And for an icon:
> 
>     * https://raw.githubusercontent.com/LMS-Community/slimserver-platforms/public/9.0/win32/res/SqueezeCenter.ico

Does the app have a webgui? or is it a server only?
If its a server only, I dont think we should add a screenshot of the client here.

stavros-k commented on ix-dev/community/lyrion-music-server/app.yaml:
```diff
@@ -0,0 +1,37 @@
+description: Lyrion Music Server - controls a wide range of Squeezebox audio players
+home: https://lyrion.org/
+host_mounts: []
+icon: https://avatars.githubusercontent.com/u/138057124?s=200&v=4
```
```suggestion
icon: https://media.sys.truenas.net/apps/lyrion-music-server/icons/icon.png
```

hartzell replied:
done

hartzell commented:
@stavros-k asked:
> Does the app have a webgui? or is it a server only?

Yes, it does, unless I mis-understand your question.  The image at:

> https://lyrion.org/assets/screenshot.png

is the main page of what LMS serves on the `web_port`, port 9000.

stavros-k commented on ix-dev/community/lyrion-music-server/app.yaml:
```diff
@@ -0,0 +1,37 @@
+  group_name: root
+  uid: 0
+  user_name: root
+sources:
```
```suggestion
screenshots:
- https://media.sys.truenas.net/apps/lyrion-music-server/screenshots/screenshot1.png
sources:
```

stavros-k commented:
I'll take a look on the cli port thingy later today and update the library. And merge it afterwards!
Nice work!

hartzell commented:
Thank you for all your help!

stavros-k commented:
@hartzell  
Looks like you have dis-allowed changes to be pushed on this PR.
I have found solutions for the cli port, can you open it so I can push them?

hartzell commented:
I'm happy to give you access, but I can't figure out how.

[This page](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) says there should be a click box on the PR, but I don't see it where they describe or anywhere else.

Can you describe to me how to do this?

stavros-k commented:
It should be on the right side of this PR 
<img width="1402" alt="image" src="https://github.com/user-attachments/assets/49b5c1a4-2108-4338-b563-e81e984ff29e" />

Under the lock conversation. Is it there?

hartzell commented:
Don't see it.  Are you officially a "maintainer" for this repo?

<img width="399" alt="Screenshot 2025-01-09 at 8 17 19 AM" src="https://github.com/user-attachments/assets/f7f9da6a-368e-462f-9dcb-542dc9958e41" />

stavros-k commented:
Yea, I was able to edit PRs on the past.
I'll send you the patch then so you can apply

stavros-k commented:
> Weird. I spend most of my work time in GItLab, so I haven't kept up with the _Hub_ UI. Sorry if I'm missing it.

No, you are not missing it. it should be there. I'll check repo settings later as well.
Sending you the patch in a bit

hartzell commented:
I might be causing it, it seems like the behavior is different if the fork belongs to an individual or an org.  I created an "org" (hartzell-forks) as a place to keep my forks so they don't get mixed in with my stuff.  Neat freak, etc...

At least that's what this PR suggests might be happening: https://github.com/orgs/community/discussions/9921

stavros-k commented:
[0001-polish-app.patch](https://github.com/user-attachments/files/18363652/0001-polish-app.patch)

hartzell commented:
I'll apply the patch and push, thank you.

Looks like it's not possible to set up, https://github.com/orgs/community/discussions/5634, even comments from other folks who use an org to manage their forks.  Sigh....

stavros-k commented:
Oh well. Now we know :D Thanks for looking for that info!

hartzell commented:
Applied.
This seems to work for me (once I adjust the test dirs)
```
.github/scripts/ci.py --app lyrion-music-server --train community --test-file basic-values.yaml
```

stavros-k commented on ix-dev/community/lyrion-music-server/app.yaml:
```diff
@@ -0,0 +1,39 @@
+- entertainment
+- music
+- streaming
+lib_version: 2.1.8
```
can you either update the library files or changed it back ot 2.1.7 and I'll update to another PR

hartzell replied:
Rebased to the tip of master, reran the `.github/... command`, git rm'd the 2.1.7 lib and added the 2.1.8 lib.

Hopefully that's what you need.

hartzell commented:
Thank you for all the help!
